misk/src/main/kotlin/misk/web/marshal/Generic.kt

```diff
@@ -0,0 +1,111 @@
+package misk.web.marshal
+
+import misk.inject.typeLiteral
+import misk.web.Response
+import misk.web.ResponseBody
+import misk.web.marshal.Marshaller.Companion.actualResponseType
+import okhttp3.MediaType
+import okio.BufferedSink
+import okio.BufferedSource
+import okio.ByteString
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Type
+import kotlin.reflect.KParameter
+import kotlin.reflect.KType
+import kotlin.reflect.jvm.javaType
+
+/** Handles generic unmarshalling, for cases where the action can accept anything */
+object GenericUnmarshallers {
+    fun canHandle(type: KType) = canHandle(type.javaType)
+    fun canHandle(type: Type) = GENERIC_REQUEST_TYPES.contains(type)
+
+    private val GENERIC_REQUEST_TYPES = setOf(
+            String::class.java,
+            Array<Byte>::class.java,
+            BufferedSource::class.java,
+            ByteString::class.java
+    )
+
+    private object ToString : Unmarshaller {
+        override fun unmarshal(source: BufferedSource): Any? = source.readUtf8()
+    }
+
+    private object ToByteArray : Unmarshaller {
+        override fun unmarshal(source: BufferedSource): Any? = source.readByteArray()
+    }
+
+    private object ToBufferedSource : Unmarshaller {
+        override fun unmarshal(source: BufferedSource): Any? = source
+    }
+
+    private object ToByteString : Unmarshaller {
+        override fun unmarshal(source: BufferedSource): Any? = source.readByteString()
+    }
+
+    fun into(parameter: KParameter): Unmarshaller {
+        val paramType = parameter.type
+        return when (paramType.typeLiteral().rawType) {
+            String::class.java -> ToString
+            Array<Byte>::class.java -> ToByteArray
```

Comment:
off topic: it’ll be really hard to explain to future developers why `Array<Byte>` is a legit raw type and `List<Byte>` is not. No action to take here.